sqlair.go

```diff
@@ -0,0 +1,230 @@
+func (db *DB) ExecContext(s *Statement, ctx context.Context, inputStructs ...any) (sql.Result, error) {
+	ce, err := s.pe.Complete(inputStructs...)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := db.db.ExecContext(ctx, ce.CompletedSQL(), ce.CompletedArgs()...)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Query takes a prepared SQLair Statement and returns a Query object for
+// iterating over the results.
+// Query uses QueryContext with context.Background internally.
+func (db *DB) Query(s *Statement, inputStructs ...any) (*Query, error) {
+	return db.QueryContext(s, context.Background(), inputStructs...)
+}
+
+// QueryContext takes a prepared SQLair Statement and returns a Query object for
+// iterating over the results.
+func (db *DB) QueryContext(s *Statement, ctx context.Context, inputStructs ...any) (*Query, error) {
+	ce, err := s.pe.Complete(inputStructs...)
+	if err != nil {
+		return nil, err
+	}
+
+	q := func() (*sql.Rows, error) {
+		return db.db.QueryContext(ctx, ce.CompletedSQL(), ce.CompletedArgs()...)
+	}
+	return &Query{outputs: ce.OutputFields(), q: q}, nil
+}
+
+// Query holds a database query and is used to iterate over the results.
+type Query struct {
+	err     error
+	rows    *sql.Rows
+	outputs expr.OutputFields
+	q       func() (*sql.Rows, error)
+}
+
+// One runs a query and decodes the first row into outputStructs.
+func (q *Query) One(outputStructs ...any) error {
+	if !q.Next() {
+		return fmt.Errorf("cannot return one row: no results")
+	}
+	q.Decode(outputStructs...)
+	return q.Close()
+}
+
+// All iterates over the query and decodes all the rows.
+// It fabricates all struct instantiations needed.
+// It returns a list of rows, each row is a list of pointers to structs.
+// The structs appear in the syntactic order they appear in the query.
+func (q *Query) All() ([][]any, error) {
+	var rows [][]any
+
+	for q.Next() {
+		cols, err := q.rows.Columns()
+		if err != nil {
+			return [][]any{}, err
+		}
+
+		ptrs, row, err := expr.FabricatedOutputAddrs(cols, q.outputs)
+		if err != nil {
+			return [][]any{}, err
+		}
+		if err := q.rows.Scan(ptrs...); err != nil {
+			return [][]any{}, err
+		}
+
+		rows = append(rows, row)
+	}
+
+	if err := q.Close(); err != nil {
+		return [][]any{}, err
+	}
+	return rows, nil
+}
+
+// Next prepares the next row for decoding.
+// The first call to Next will execute the query.
+// If an error occurs it will be returned with Query.Close().
+func (q *Query) Next() bool {
+	if q.err != nil {
+		return false
+	}
+	if q.rows == nil {
+		rows, err := q.q()
+		if err != nil {
+			q.err = err
+			return false
+		}
+		q.rows = rows
+	}
+	return q.rows.Next()
+}
+
+// Decode decodes the current result into the structs in dests.
+// dests must contain all the structs mentioned in the query.
+// If an error occurs it will be returned with Query.Close().
+func (q *Query) Decode(dests ...any) (ok bool) {
```

Comment:
`outputStructs` from `One` feels nicer. In either case, we should use matching terms.